### PR TITLE
Add SmolVLA kernel Programs with golden data and MLIR

### LINKED ISSUE #36: [Kernels] [SmolVLA] [Coverage] Main issue to track kernel coverage for SmolVLA
# SmolVLA NPU Kernel Implementation

## Summary

We need to implement NPU kernels for the SmolVLA model (PI0 policy). **4 kernel types cover 99.8% of compute.** This PR tracks progress across all kernels, each as a subissue if possible.

An example kernel (SiLU) is implemented and tested in `npu_model`  which can be used as a template.

> **Note on fp8**: Currently weights are stored in MX fp8 but activations remain bf16. We want full fp8 compute (activations too). Design kernels with fp8 in mind. We'll sync with the quantization team, start with what's in the MLIR now and we adjust later.

## Kernel Progress

> After creating this issue, create a subissue for each kernel below. Replace `<!-- #N -->` with the subissue number (e.g., `#42`). When a subissue is closed, the checkbox auto-updates.

### P0 — Critical compute (98.4% of FLOPs)

- [ ]  `quantized_matmul_fp8` — bf16 act x fp8 weight + block scale (379 instances, 82.2%) — 10 shape variants
- [ ]  `fused_attention` — SDPA: `Q@K + softmax + @V` (36 instances, 16.1%) — 1 shape variant

### P1 — Attention path + int8 projections (1.4%)

- [ ]  `batch_matmul` — attention score/value matmuls (46 instances, 0.6%) — 6 shape variants
- [ ]  `softmax` — linalg.softmax (23 instances) — 3 shape variants
- [ ]  `matmul_i8` — int8 dynamic-quantized projections (67 instances, 0.8%) — 6 shape variants

### P2 — Activations, norms, quantization

- [ ]  `rms_norm` — reduce + rsqrt + scale (139 instances) — 3 shape variants
- [ ]  `gelu_tanh` — GELU activation (36 instances) — 1 shape variant
- [x] `silu` — SiLU/Swish activation (32 instances) — 3 shape variants — done: `npu_model/configs/programs/smolvla_silu.py`
- [ ]  `int8_quantize` — dynamic bf16 to i8 (34 instances) — 1 shape variant
- [ ]  `rope` — rotary position embedding (138 instances) — 2 shape variants

### P3 — Elementwise, reductions, type casts

- [ ]  `elementwise_add` (539 instances) — 16 shape variants
- [ ]  `elementwise_mul` (664 instances) — 19 shape variants
- [ ]  `elementwise_div` (217 instances) — 6 shape variants
- [ ]  `elementwise_sub` (114 instances) — 6 shape variants
- [ ]  `reduction_sum` (214 instances) — 3 shape variants
- [ ]  `type_conversion` — bf16/f32 casts (655 instances) — 49 shape variants
- [ ]  `bias_add_cast` (75 instances) — 1 shape variant

### Fused composite kernels

These ops always appear consecutively and can be fused into a single dispatch:

- [ ]  `fused_matmul_bias` — quantized_matmul + elementwise_add (283x)
- [ ]  `fused_quant_matmul` — int8_quantize + linalg.matmul (30x)
- [ ]  `fused_silu_gate` — silu + elementwise_mul (32x)
- [ ]  `fused_norm_scale` — rms_norm + elementwise_mul (64x)

Progress: 1 / 21 kernels (+ 0 / 4 fused)

## How to implement a kernel (step by step)

Reference example: `npu_model/configs/programs/smolvla_silu.py`

### 1. Generate the analysis data

The MLIR kernel files and golden data are **generated locally** (gitignored). Run from the merlin repo root:

```bash
# Compile the model to global-opt (vanilla, no accelerator plugins)
conda run -n merlin-dev uv run tools/merlin.py compile \
  models/smolVLA/smolVLA.q.fp8.mlir \
  --target spacemit_x60 --quantized \
  --compile-to global-optimization --dump-phases

# Strip weight blobs so MLIR files are readable
conda run -n merlin-dev uv run tools/strip_mlir_weights.py \
  build/compiled_models/smolVLA/spacemit_x60_RVV_smolVLA.q.fp8/ --in-place

# Generate kernel MLIR catalog (standalone .mlir per shape variant, resolved affine maps)
conda run -n merlin-dev uv run benchmarks/SaturnNPU/scripts/extract_all_kernel_variants.py

# Generate golden test data (PyTorch + IREE cross-check)
conda run -n merlin-dev uv run benchmarks/SaturnNPU/scripts/generate_npu_golden_tests.py \
  --output-dir benchmarks/SaturnNPU/golden_data/ --scale both
conda run -n merlin-dev uv run benchmarks/SaturnNPU/scripts/generate_mlir_golden_data.py \
  --output-dir benchmarks/SaturnNPU/golden_data/mlir_level/
conda run -n merlin-dev uv run benchmarks/SaturnNPU/scripts/export_golden_data.py \
  benchmarks/SaturnNPU/golden_data/ --formats numpy
```

After running, you'll have locally:

```text
benchmarks/SaturnNPU/
  kernels/<kernel_type>/           -- standalone .mlir files per shape variant
  golden_data/small/<layer>/       -- PyTorch golden data (.pt + .npy)
  golden_data/mlir_level/<op>/     -- IREE-compiled golden data (.npy)
```

Full instructions: `benchmarks/SaturnNPU/README.md`

### 2. Understand what the kernel does

Open `kernels/<kernel_type>/` — each `.mlir` file shows the full `linalg.generic` body with types, indexing maps, and iterator types. Affine maps are fully resolved (no `#mapN` references — extracted via MLIR Python bindings).

For fused kernels, the directory has `op_a.mlir` and `op_b.mlir` showing the two ops that chain together.

### 3. Get golden data

**PyTorch-level** (per-layer, with operator decomposition):

```text
benchmarks/SaturnNPU/golden_data/small/<layer_type>/
  layer_input_x.pt             -- layer input
  layer_output_output.pt       -- expected layer output
  operators/
    00_<op>/input_*.pt          -- operator input
    00_<op>/output_*.pt         -- operator output (feeds into next)
```

Available in `.npy` (NumPy) via `golden_data/small_numpy/`.

**MLIR-level** (compiled via IREE, per-op):

```text
benchmarks/SaturnNPU/golden_data/mlir_level/<op_variant>/
  input_0.npy, input_1.npy     -- op inputs
  output_iree.npy              -- output from IREE compiler (ground truth)
  metadata.json
```

You can also generate golden data inline using the IREE Python bindings -> see the SiLU example which compiles and runs its MLIR definition directly in the test file.

**Real model weights**: The golden data above uses random weights (deterministic seed). The actual model weights are embedded in the unstripped MLIR (`models/smolVLA/smolVLA.q.fp8.mlir`, ~886MB). For initial kernel development, random-weight golden data should be sufficient.

### 4. Write your kernel program

Copy `npu_model/configs/programs/smolvla_silu.py` and replace:

```python
# 1. MLIR_DEFINITION -- paste the linalg.generic from kernels/<type>/variant_*.mlir
#    (use f32 types so IREE can compile it on any CPU)

# 2. reference_fn(x) -- PyTorch function that computes the expected output

# 3. ISA instructions -- map each MLIR body op to NPU instructions:
#    arith.negf   -> vmul.bf16(x, -1)
#    math.exp     -> vexp.bf16
#    arith.addf   -> vadd.bf16
#    arith.divf   -> vrecip.bf16 + vmul.bf16
#    math.rsqrt   -> vsqrt.bf16 + vrecip.bf16
#    math.tanh    -> vtanh.bf16
#    ... (see isa_definition.py for all available ops)

# 4. golden_result -- tuple of (dram_output_address, expected_tensor)
```

The file contains the MLIR definition, the PyTorch reference, and the ISA program all in one place. If `iree.compiler` is installed, it cross-checks MLIR vs PyTorch automatically.

### 5. Place the file

```text
third_party/npu_model/npu_model/configs/programs/smolvla_<kernel_name>.py
```

No other files need to change. The `__init__.py` uses `import_packages()` which auto-discovers any new `Program` subclass.

Shared PyTorch references go in `npu_model/workload/smolvla_ops.py`.

### 6. Test locally

```bash
cd third_party/npu_model
uv run python scripts/test_programs.py --verbose --max-cycles 100000
```

Your program appears as `OK SmolVLA<Name>Program` if the golden check passes.

## How to create a subissue

For each kernel you work on, create a subissue of this tracking issue with:

- **Title**: `Kernel: <kernel_name>` (e.g., `Kernel: gelu_tanh`)
- **Body**:
  - Which shape variant(s) you're implementing
  - The MLIR body (copy from the generated `.mlir` file)
  - Your ISA mapping plan
- **Done when**: `smolvla_<name>.py` added, `test_programs.py` passes, PR merged

## Model architecture

```text
SmolVLA (lerobot/smolvla_base)
+-- SigLIP Vision Tower -- 12 encoder layers x 3 images
|   +-- Attention: LayerNorm -> Q/K/V proj (fp8 matmul) -> SDPA -> O proj
|   +-- MLP: LayerNorm -> FC1 (fp8 matmul) -> GELU -> FC2 (fp8 matmul)
+-- Gemma Main LM -- decoder layers
|   +-- Attention: RMSNorm -> Q/K/V (fp8) -> RoPE -> score -> softmax -> value -> O
|   +-- MLP: RMSNorm -> gate/up (fp8) -> SiLU -> gate*up -> down (fp8)
+-- Gemma Expert -- decoder layers (int8 fallback: hidden_dim=720, 720%32!=0)
|   +-- Same structure, cross-attention variant (Q[50] x KV[241])
+-- Projections: action_in/out, state, action_time_mlp
```

## Quantization status

| Component | Weights | Activations | Note |
| --- | --- | --- | --- |
| SigLIP + Gemma main | MX fp8 (f8E4M3FN) | bf16 | block_size=32 |
| Gemma expert | int8 (fallback) | bf16 | 720 % 32 != 0 |
| Attention Q@K, attn@V | n/a | bf16/f32 | activation x activation |
| Softmax, norms | n/a | bf16/f32 | numerical stability |

TODO (quantization team): (1) get expert layers to fp8 (requires TorchAO changes for non-32-aligned dims), (2) explore fp8 activation quantization for compute, not just storage.

## Reference

| Resource | Location | 
| --- | --- | 
| Analysis scripts | `benchmarks/SaturnNPU/scripts/` | 
| Setup instructions | `benchmarks/SaturnNPU/README.md` | 
| SiLU example kernel | `npu_model/configs/programs/smolvla_silu.py` | 
| Kernel MLIR files | `benchmarks/SaturnNPU/kernels/` | 
| Golden test data | `benchmarks/SaturnNPU/golden_data/` | 
| Layer decomposition trace | `benchmarks/SaturnNPU/LAYER_DECOMPOSITION_TRACE.md` | 
| Sankey diagram | `benchmarks/SaturnNPU/plots/sankey_decomposition.html` | 
| Vanilla global-opt MLIR | `build/.../spacemit_x60_.../module.4.global-optimization.mlir` |

### LINKED ISSUE #43: Kernel: fused_matmul_bias
[smolvla_fused_matmul_bias.py](https://github.com/user-attachments/files/26692658/smolvla_fused_matmul_bias.py)

### LINKED ISSUE #44: Kernel: fused_norm_scale
[smolvla_fused_norm_scale.py](https://github.com/user-attachments/files/26692699/smolvla_fused_norm_scale.py)

### LINKED ISSUE #45: Kernel: fused_silu_gate
[smolvla_fused_silu_gate.py](https://github.com/user-attachments/files/26692705/smolvla_fused_silu_gate.py)

### PR DESCRIPTION
**Pull request by copparihollmann**
List of kernels added with Golden data and MLIR implementation:
- silu (pre-existing template, unchanged)
- elementwise_add, elementwise_mul, elementwise_sub, elementwise_div
- matmul (single-tile 32x32) and matmul_k_chain (K=64, two-tile accumulate)
- rms_norm, softmax, reduction_sum, requant
- rope_frequency, gelu_tanh
- attention (single-tile SDPA) and fused_attention (flash attention, KV=2)

Each Program is self-contained (MLIR string + PyTorch reference + seeded golden tensor + inline ISA) and validated two ways at import time / via test_programs.py: simulator output vs golden, and IREE compile+run of the MLIR vs PyTorch reference.

Changes outside npu_model/configs/programs/:
- npu_model/hardware/arch_state.py: byte-level MRF fp8 write (view as uint8 to avoid value-cast) and 0-byte DMA write bounds-check.
- scripts/test_programs.py: per-program kernel_tolerance override so fp8-quantized kernels (e.g. attention) can opt into a looser tol.
- npu_model/configs/isa_definition.py -> vpack.bf16.fp8 now writes the quantized tile as uint8 bytes, pairing with the arch_state fp8 byte-copy fix.

Powered by Autocomp lol

### CODE REVIEW

**Comment by copparihollmann** [comment]
Will close #36 

Still some todo with the tests that Lux mentioned during todays meeting

**Comment by copparihollmann** [comment]
This will also close #43 , #44 , and #45 .

@Jeremy-Ryan-Mills I added you in the commit where I added your kernels and used claude to get some help to port them based on what @reed-nicolas sent me

**Review by iansseijelly — APPROVED** [review]
LGTM
